File: app/src/main/java/com/kidozh/discuzhub/results/UserNoteListResult.java
package com.kidozh.discuzhub.results;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.kidozh.discuzhub.entities.ThreadInfo;
import com.kidozh.discuzhub.utilities.OneZeroBooleanJsonDeserializer;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@JsonIgnoreProperties(ignoreUnknown = true)
public class UserNoteListResult extends BaseResult {
    @JsonProperty("Variables")
    public NoteListVariableResult noteListVariableResult;

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class NoteListVariableResult extends VariableResults{

        @JsonProperty("list")
        public List<UserNotification> notificationList;
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int count, page;
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonDeserialize(using = OneZeroBooleanJsonDeserializer.class)
        public int perPage;

    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class UserNotification{
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int id, uid;
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonProperty("authorid")
        public int authorId;
        public String type, note, author="";
        @JsonProperty("new")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonDeserialize(using = OneZeroBooleanJsonDeserializer.class)
        public boolean isNew;
        @JsonProperty("dateline")
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "s")
        public Date date;
        @JsonProperty("from_id")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int fromId;
        @JsonProperty("from_idtype")
        public String fromIdType;
        @JsonProperty("from_num")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int fromNum;
        @JsonProperty("notevar")
        @JsonIgnoreProperties(ignoreUnknown = true)
        public UserNotificationExtraInfo notificationExtraInfo;

    }

    public static class UserNotificationExtraInfo{
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int tid, pid;
        public String subject;
        @JsonProperty("actoruid")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        public int actorUid;
        @JsonProperty("actorusername")
        public String actorUsername;
    }

    public boolean isError(){
        return this.message == null;
    }
}

File: app/src/main/java/com/kidozh/discuzhub/results/UserNoteListResult.kt
package com.kidozh.discuzhub.results

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.kidozh.discuzhub.utilities.OneZeroBooleanJsonDeserializer
import java.util.*
import kotlin.collections.ArrayList

@JsonIgnoreProperties(ignoreUnknown = true)
class UserNoteListResult : BaseResult() {
    @JvmField
    @JsonProperty("Variables")
    var noteListVariableResult: NoteListVariableResult = NoteListVariableResult()

    @JsonIgnoreProperties(ignoreUnknown = true)
    class NoteListVariableResult : VariableResults() {
        @JvmField
        @JsonProperty("list")
        var notificationList: List<UserNotification> = ArrayList()

        @JvmField
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var count = 0

        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var page = 0

        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonDeserialize(using = OneZeroBooleanJsonDeserializer::class)
        var perPage = 0
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    class UserNotification {
        @JvmField
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var id = 0

        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var uid = 0

        @JvmField
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonProperty("authorid")
        var authorId = 0
        @JvmField
        var type: String = ""
        @JvmField
        var note: String = ""
        @JvmField
        var author = ""

        @JvmField
        @JsonProperty("new")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        @JsonDeserialize(using = OneZeroBooleanJsonDeserializer::class)
        var isNew = false

        @JvmField
        @JsonProperty("dateline")
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "s")
        var date: Date = Date()

        @JsonProperty("from_id")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var fromId = 0

        @JsonProperty("from_idtype")
        var fromIdType: String = ""

        @JsonProperty("from_num")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var fromNum = 0

        @JvmField
        @JsonProperty("notevar")
        @JsonIgnoreProperties(ignoreUnknown = true)
        var notificationExtraInfo: UserNotificationExtraInfo? = null
    }

    class UserNotificationExtraInfo {
        @JvmField
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var tid = 0

        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var pid = 0
        @JvmField
        var subject: String = ""

        @JsonProperty("actoruid")
        @JsonFormat(shape = JsonFormat.Shape.STRING)
        var actorUid = 0

        @JsonProperty("actorusername")
        var actorUsername: String? = null
    }

    override fun isError(): Boolean {
        return message == null
    }
}